crates/factor-outbound-pg/src/host.rs

```diff
@@ -0,0 +1,423 @@
+use anyhow::{anyhow, Result};
+use native_tls::TlsConnector;
+use postgres_native_tls::MakeTlsConnector;
+use spin_core::{async_trait, wasmtime::component::Resource};
+use spin_world::v1::postgres as v1;
+use spin_world::v1::rdbms_types as v1_types;
+use spin_world::v2::postgres::{self as v2, Connection};
+use spin_world::v2::rdbms_types;
+use spin_world::v2::rdbms_types::{Column, DbDataType, DbValue, ParameterValue, RowSet};
+use tokio_postgres::{
+    config::SslMode,
+    types::{ToSql, Type},
+    Client, NoTls, Row, Socket,
+};
+use tracing::instrument;
+use tracing::Level;
+
+use crate::InstanceState;
+
+impl InstanceState {
+    async fn open_connection(&mut self, address: &str) -> Result<Resource<Connection>, v2::Error> {
+        self.connections
+            .push(
+                build_client(address)
+                    .await
+                    .map_err(|e| v2::Error::ConnectionFailed(format!("{e:?}")))?,
+            )
+            .map_err(|_| v2::Error::ConnectionFailed("too many connections".into()))
+            .map(Resource::new_own)
+    }
+
+    async fn get_client(&mut self, connection: Resource<Connection>) -> Result<&Client, v2::Error> {
+        self.connections
+            .get(connection.rep())
+            .ok_or_else(|| v2::Error::ConnectionFailed("no connection found".into()))
+    }
+
+    async fn is_address_allowed(&self, address: &str) -> bool {
+        let Ok(config) = address.parse::<tokio_postgres::Config>() else {
+            return false;
+        };
+        for (i, host) in config.get_hosts().iter().enumerate() {
+            match host {
+                tokio_postgres::config::Host::Tcp(address) => {
+                    let ports = config.get_ports();
+                    // The port we use is either:
+                    // * The port at the same index as the host
+                    // * The first port if there is only one port
+                    let port =
+                        ports
+                            .get(i)
+                            .or_else(|| if ports.len() == 1 { ports.get(1) } else { None });
+                    let port_str = port.map(|p| format!(":{}", p)).unwrap_or_default();
+                    let url = format!("{address}{port_str}");
+                    // TODO: Should I be unwrapping this?
```

Comment:
Not sure if this is copied from the current implementation, but it does seem like `is_address_allowed` should be returning an `anyhow::Result<bool>` so that this unwrap isn't necessary.